Content.Client/Light/RgbLightControllerSystem.cs

```diff
@@ -1,40 +1,96 @@
 using System;
+using Content.Shared.Item;
 using Content.Shared.Light;
 using Content.Shared.Light.Component;
 using Robust.Client.GameObjects;
 using Robust.Shared.GameObjects;
+using Robust.Shared.GameStates;
 using Robust.Shared.IoC;
 using Robust.Shared.Maths;
 using Robust.Shared.Timing;
 
 namespace Content.Client.Light
 {
-    public class RgbLightControllerSystem : EntitySystem
+    public sealed class RgbLightControllerSystem : SharedRgbLightControllerSystem
     {
         [Dependency] private IGameTiming _gameTiming = default!;
 
         public override void Initialize()
         {
             base.Initialize();
 
-            UpdatesOutsidePrediction = true;
+            SubscribeLocalEvent<RgbLightControllerComponent, ComponentHandleState>(OnHandleState);
+            SubscribeLocalEvent<RgbLightControllerComponent, ComponentRemove>(OnComponentRemoved);
         }
 
-        public override void Update(float frameTime)
+        private void OnComponentRemoved(EntityUid uid, RgbLightControllerComponent rgb, ComponentRemove args)
+        {
+            if (TryComp(uid, out PointLightComponent? light))
+                light.Color = Color.White;
+
+            if (TryComp(uid, out SharedItemComponent? item))
+                item.Color = Color.White;
+
+            ResetSpriteColors(uid, rgb);
+        }
+
+        private void OnHandleState(EntityUid uid, RgbLightControllerComponent rgb, ref ComponentHandleState args)
+        {
+            if (args.Current is not RgbLightControllerState state)
+                return;
+
+            // just to be safe, un-colour the layers so they don't get stuck with some mid-transition one.
+            ResetSpriteColors(uid, rgb);
+
+            rgb.CycleRate = state.CycleRate;
+            rgb.Layers = state.Layers;
+        }
+
+        private void ResetSpriteColors(EntityUid uid, RgbLightControllerComponent? rgb = null, SpriteComponent? sprite = null)
+        {
+            if (!Resolve(uid, ref rgb, ref sprite))
+                return;
+
+            if (rgb.Layers == null)
+            {
+                sprite.Color = Color.White;
+                return;
+            }
+
+            foreach (var layer in rgb.Layers)
+            {
+                sprite.LayerSetColor(layer, Color.White);
+            }
+        }
```

Comment:
This assumes the sprite started off with no coloration. Wouldn't this be incorrect for eswords?

Reply:
Instead of adding several esword sprites, the colours can just be set by the same sprite colour field which was what was/is now being done in the esword PR. Also without grey-scale sprites using rgb controller for the rainbow sword wouldn't really work.

But yes, in general this should probably cache the colors from before the rgb was enabled and revert it properly once disabled. I'll add that functionality.

Reply:
Yes it would, I'll add a `OriginalColor` fields and properly restore the pre-rgb values